### DISCUSSION TITLE
Form模型能否内置支持扩展属性

### DISCUSSION THREAD

**Discussion by rxlf** [alibaba/formily#2431] [Ideas]
类似 `Field` 模型的 data 属性，把一些全局的数据&状态挂载到 Form 扩展属性上，能很大程度上简化使用

一些场景化例子

eg:
将全局样式&属性挂载到 form 上，在自定义组件中去消费

eg2:
在表单初始化前去拉取全局数据，然后挂载到表单上，field 里取消费数据

eg3:
表单的状态，如是否为编辑态 isEdit ，在组件中去消费这个状态展示不同的样式

**Comment by janryWang** [comment]
如果只是自定义组件内部消费，为啥不直接创建context下发下去

**Reply by rxlf** [reply]
可能我没描述清楚，不仅仅是需要在自定义组件中去消费，在表单的任何场景中（组件、装饰器、 Effect ）都有可能用到，同时需要使用响应式的能力，避免业务手动的去维护状态，能极大简化开发成本：

针对 Field, 我们有这样的场景， 多个字段依赖同一个接口下发数据源，但是接口存在失败的情况，需要提示用户重试，这时候我可以把请求的状态存到 field.data 里（如失败的状态、哪些字段需要重新请求），本身支持的 field.data 就能很大程度上简化流程。 

对 Form 也是一样的，存在一些全局状态&属性，这些全局状态&属性在组件、装饰器、 Effect 钩子里，都是会使用到的，同时本身支持响应式的话，所以要是 Form 层能有支持的属性，会极大的减小开发成本

**Reply by janryWang** [reply]
还是没明白，你自己顶层管理的observable数据，react context下发下去，所有地方都能消费的呀，为什么一定要耦合到form上呢？field的data属性存在的意义主要是因为扩展状态与字段的映射关系手动维护起来很麻烦才支持的，如果简单的话，字段data，我也是推荐用户自己外部单独管理

**Reply by kaoding** [reply]
不建议支持这一属性，这种需求有很多方法可以解决，框架能简单的就简单。

**Comment by heaven2049** [comment]
使用Context可以解决，但繁杂，且每个组件都需要明确感知Context的存在，但实际并不想每个组件都去新开发一次 & 消费Context。
例子1，Input组件、RadioButton组件要同时支持编辑态、阅读态，编辑态会直接使用antd/arco的组件，阅读态使用mapProps+readPretty封装，使用时传 readPretty 即可，我们在ArcoForm里会将 readPretty 支持配置为响应式函数，依赖全局属性（如isEdit）做响应式（如创建表单编辑态，编辑表单阅读态），而非开发一个新的组件Input、RadioButton组件去消费Context。
例子2，我们发现表单的场景中，字段的初始化、字段的联动是极其常见的场景。故结合Formily的effects，开发了若干helper工具，比如 同步初始化， syncInit(pattern, callback(field, form) => {} ) {} 。 此时，如若字段需要消费全局内容，直接消费form.data即可而且为响应式，可以很方便的写纯逻辑，而不用和UI组件打交道。

**Reply by janryWang** [reply]
不需要重新开发组件去消费context的，可以参考mapProps和mapReadpretty，都是无侵入的包装函数，你们也可以通过这样来消费顶层传下来的私有状态

**Reply by heaven2049** [reply]
感谢解答，确实可以通过mapProps实现该不侵入组件，而实现控制的行为。但该方式的问题在于，和React组件强绑定，且全局数据和Form的effects混搭时，因为不是一个体系，两者的结合比较费劲。
我们目前倾向利用Form的effects+响应式的体系，去简化Form中逻辑拼装的成本，能用effects解决的场景，尽量不使用自定义组件。
再举一些effects中需要全局数据的实际的例子，比如多字段dataSource依赖同一接口返回，比如表单创建态/编辑态转换。我们目前是扩充了表单生命周期，加了一个同步/异步钩子用于加载表单的全局数据（响应式），当多字段依赖同一数据源接口时，只需写多个effects并消费全局数据皆可，无需在React层做处理。（实际上这边对effects的使用也有封装，来进一步简化使用成本）。表单状态切换同理，对于组件只需设置响应阅读态，但字段的阅读态可由effects读取全局数据的值做简易处理。尤其当此时有其它effects要一起生效时（实际上并不会只单单切换阅读态这么简单），也方便组织，因为它们都属于Formily的副作用及响应式体系内。
最后，目前该特性其实我们在字节内的封装已经Hack实现了，也论证了确实在业务中有简化使用成本的帮助。不过Hack的封装带来了一些边际成本，我们希望能绕过去。

**Reply by heaven2049** [reply]
可能前面的描述有误导，全局数据其实我们想兼容两个场景，1. 自定义组件。2.effects内使用，并且满足两个场景的混搭。前面花了很多精力讲组件内使用，应该是我们讨论描述不太准确，抱歉。

**Reply by janryWang** [reply]
所以你们就是单纯想扩展模型，而且希望做到跨框架使用是吧？

**Reply by heaven2049** [reply]
是的，我们抽象中的一层是Formily场景化增强，这一层是跨框架的。

**Reply by janryWang** [reply]
我想了下，感觉还是在mapProps这一层处理比较好，因为即便你扩展了表单模型，但组件层，或者组件适配层始终还是得感知这种扩展属性，然后转换到组件可以识别的属性，所以扩展表单模型的意义就没那么大了，反而还会增加理解成本

**Reply by rxlf** [reply]
在 mapProps 怎么去获取顶层 Form 的属性呢，或者说 Form 全局属性，保存到哪里比较合适呢

**Reply by janryWang** [reply]
我的意思是，你们单独用@formily/reactive定义一个面向你们场景的observable模型，然后通过context下发，在mapProps里直接useContext消费就行了

**Reply by heaven2049** [reply]
其实目前已经这么做了，只是没用context下发，而是我们直接挂在了Form上，方便effects中取，组件中也直接用 obversable 组件 +useForm 来消费。

**Reply by heaven2049** [reply]
组件感知类型可以通过给form.data支持泛型来感知，这一层是不可避免的，但这一层不是百分比必须的，因为许多逻辑是通过effects来组装的。
但假如想在effects中结合使用（我们的一个场景），就得必须扩展表单模型了。现在我们得在Form创建完成时挂载一个 formily/reactive 的observable 对象，和字段初始化等缠绕在一起，总体有略多Hack成本。

**Reply by janryWang** [reply]
要在effects消费是可以用createEffectContext的方式提供个hook给用户使用的

**Reply by heaven2049** [reply]
OK，我们调研下这个钩子看看

**Comment by shubinla** [comment]
如果是依赖某些全局的数据，比如引用的 store 数据，想要在 json schema 中去引用，有什么好的方法吗，目前我是通过挂在一个 field 的scope 上面，然后在 json schema 中去引用 {{xxx}}。目前项目中比较多依赖的数据，需要在 schema 中的很多属性用到，比如 visible, reaction, validator 等等。
